### ISSUE TITLE
Crash when executing PyTorch-trained YoloV3 on NCS2/MyriadX

### ISSUE THREAD

**Issue by mhkabir** [openvinotoolkit/openvino#1480]
OpenVino: 2020.4
OS: Ubuntu 18.04, kernel 5.4.0-42-generic

We have a YoloV3 model trained using PyTorch, and then exported to ONNX and finally IR XML. We have run this model successfully on CPU and GPU, but execution fails on NCS2/HDDL devices. Other networks, including pretrained Darknet->TF->IR YoloV3 runs fine on the NCS2 using benchmark_app, so it's likely not a power/hardware/setup issue.

To create a minimal, reproducible example, we're using `benchmark_app`:

```
./benchmark_app -d MYRIAD -m model.xml

[Step 1/11] Parsing and validating input arguments
[ INFO ] Parsing input parameters
[Step 2/11] Loading Inference Engine
[ INFO ] InferenceEngine: 
	API version ............ 2.1
	Build .................. 2020.4.0-359-21e092122f4-releases/2020/4
	Description ....... API
[ INFO ] Device info: 
	MYRIAD
	myriadPlugin version ......... 2.1
	Build ........... 2020.4.0-359-21e092122f4-releases/2020/4

[Step 3/11] Setting device configuration
[Step 4/11] Reading the Intermediate Representation network
[ INFO ] Loading network files
[ INFO ] Read network took 122.67 ms
[Step 5/11] Resizing network to match image sizes and given batch
[ INFO ] Network batch size: 1
[Step 6/11] Configuring input of the model
[Step 7/11] Loading the model to the device
[ INFO ] Load network took 4331.13 ms
[Step 8/11] Setting optimal runtime parameters
[Step 9/11] Creating infer requests and filling input blobs with images
[ INFO ] Network input 'input.1' precision U8, dimensions (NCHW): 1 3 416 416 
[ WARNING ] No input files were given: all inputs will be filled with random values!
[ INFO ] Infer Request 0 filling
[ INFO ] Fill input 'input.1' with random values (image is expected)
[ INFO ] Infer Request 1 filling
[ INFO ] Fill input 'input.1' with random values (image is expected)
[ INFO ] Infer Request 2 filling
[ INFO ] Fill input 'input.1' with random values (image is expected)
[ INFO ] Infer Request 3 filling
[ INFO ] Fill input 'input.1' with random values (image is expected)
[Step 10/11] Measuring performance (Start inference asyncronously, 4 inference requests, limits: 60000 ms duration)
E: [global] [    540485] [benchmark_app] addEvent:264	Condition failed: event->header.flags.bitField.ack != 1
E: [global] [    540485] [benchmark_app] addEventWithPerf:276	 addEvent(event) method call failed with an error: 3
E: [global] [    540485] [benchmark_app] XLinkReadData:156	Condition failed: (addEventWithPerf(&event, &opTime))
E: [ncAPI] [    540485] [benchmark_app] ncFifoReadElem:3280	Packet reading is failed.
E: [global] [    540485] [Scheduler00Thr] dispatcherEventSend:53	Write failed (header) (err -4) | event XLINK_WRITE_REQ

E: [xLink] [    540485] [Scheduler00Thr] sendEvents:1027	Event sending failed
E: [global] [    540485] [Scheduler00Thr] dispatcherEventSend:53	Write failed (header) (err -4) | event XLINK_WRITE_REQ

E: [xLink] [    540485] [Scheduler00Thr] sendEvents:1027	Event sending failed
E: [global] [    540486] [benchmark_app] addEvent:264	Condition failed: event->header.flags.bitField.ack != 1
E: [global] [    540486] [benchmark_app] addEventWithPerf:276	 addEvent(event) method call failed with an error: 3
E: [global] [    540486] [benchmark_app] XLinkReadData:156	Condition failed: (addEventWithPerf(&event, &opTime))
E: [ncAPI] [    540486] [benchmark_app] getGraphMonitorResponseValue:1903	XLink error, rc: X_LINK_ERROR
E: [ncAPI] [    540486] [benchmark_app] ncGraphQueueInference:3443	Can't get trigger response
E: [global] [    540486] [Scheduler00Thr] dispatcherEventSend:53	Write failed (header) (err -4) | event XLINK_WRITE_REQ

E: [xLink] [    540486] [Scheduler00Thr] sendEvents:1027	Event sending failed
E: [global] [    540486] [Scheduler00Thr] dispatcherEventSend:53	Write failed (header) (err -4) | event XLINK_WRITE_REQ

E: [xLink] [    540486] [Scheduler00Thr] sendEvents:1027	Event sending failed
E: [global] [    540486] [benchmark_app] addEvent:264	Condition failed: event->header.flags.bitField.ack != 1
E: [global] [    540486] [benchmark_app] addEventWithPerf:276	 addEvent(event) method call failed with an error: 3
E: [global] [    540486] [benchmark_app] XLinkReadData:156	Condition failed: (addEventWithPerf(&event, &opTime))
E: [ncAPI] [    540486] [benchmark_app] getGraphMonitorResponseValue:1903	XLink error, rc: X_LINK_ERROR
E: [ncAPI] [    540486] [benchmark_app] ncGraphQueueInference:3443	Can't get trigger response
E: [global] [    540486] [Scheduler00Thr] dispatcherEventSend:53	Write failed (header) (err -4) | event XLINK_WRITE_REQ

E: [xLink] [    540486] [Scheduler00Thr] sendEvents:1027	Event sending failed
E: [global] [    540486] [Scheduler00Thr] dispatcherEventSend:53	Write failed (header) (err -4) | event XLINK_WRITE_REQ

E: [xLink] [    540486] [Scheduler00Thr] sendEvents:1027	Event sending failed
E: [global] [    540486] [benchmark_app] addEvent:264	Condition failed: event->header.flags.bitField.ack != 1
E: [global] [    540486] [benchmark_app] addEventWithPerf:276	 addEvent(event) method call failed with an error: 3
E: [global] [    540486] [benchmark_app] XLinkReadData:156	Condition failed: (addEventWithPerf(&event, &opTime))
E: [ncAPI] [    540486] [benchmark_app] getGraphMonitorResponseValue:1903	XLink error, rc: X_LINK_ERROR
E: [ncAPI] [    540486] [benchmark_app] ncGraphQueueInference:3443	Can't get trigger response
W: [global] [    540487] [Scheduler00Thr] dispatcherLocalEventGetResponse:132	Blocked event would cause dispatching thread to wait on semaphore infinitely

E: [global] [    540487] [Scheduler00Thr] dispatcherEventSend:53	Write failed (header) (err -4) | event XLINK_WRITE_REQ

E: [xLink] [    540487] [Scheduler00Thr] sendEvents:1027	Event sending failed
E: [global] [    540487] [benchmark_app] addEvent:264	Condition failed: event->header.flags.bitField.ack != 1
E: [global] [    540487] [benchmark_app] addEventWithPerf:276	 addEvent(event) method call failed with an error: 3
E: [global] [    540487] [benchmark_app] XLinkReadData:156	Condition failed: (addEventWithPerf(&event, &opTime))
E: [ncAPI] [    540487] [benchmark_app] getGraphMonitorResponseValue:1903	XLink error, rc: X_LINK_ERROR
W: [ncAPI] [    540487] [benchmark_app] ncFifoDestroy:3148	myriad NACK

[Warning][VPU][MyriadPlugin] ncFifoDelete result NC_ERROR
E: [global] [    540487] [Scheduler00Thr] dispatcherEventSend:53	Write failed (header) (err -4) | event XLINK_WRITE_REQ

E: [xLink] [    540487] [Scheduler00Thr] sendEvents:1027	Event sending failed
E: [global] [    540487] [benchmark_app] addEvent:264	Condition failed: event->header.flags.bitField.ack != 1
E: [global] [    540487] [benchmark_app] addEventWithPerf:276	 addEvent(event) method call failed with an error: 3
E: [global] [    540487] [benchmark_app] XLinkReadData:156	Condition failed: (addEventWithPerf(&event, &opTime))
E: [ncAPI] [    540487] [benchmark_app] getGraphMonitorResponseValue:1903	XLink error, rc: X_LINK_ERROR
W: [ncAPI] [    540487] [benchmark_app] ncFifoDestroy:3148	myriad NACK

[Warning][VPU][MyriadPlugin] ncFifoDelete result NC_ERROR
E: [global] [    540487] [Scheduler00Thr] dispatcherEventSend:53	Write failed (header) (err -4) | event XLINK_WRITE_REQ

E: [xLink] [    540487] [Scheduler00Thr] sendEvents:1027	Event sending failed
E: [global] [    540487] [benchmark_app] addEvent:264	Condition failed: event->header.flags.bitField.ack != 1
E: [global] [    540487] [benchmark_app] addEventWithPerf:276	 addEvent(event) method call failed with an error: 3
E: [global] [    540487] [benchmark_app] XLinkReadData:156	Condition failed: (addEventWithPerf(&event, &opTime))
E: [ncAPI] [    540487] [benchmark_app] getGraphMonitorResponseValue:1903	XLink error, rc: X_LINK_ERROR
[Warning][VPU][MyriadPlugin] Deallocate Graph result NC_ERROR.
W: [ncAPI] [    540493] [benchmark_app] ncDeviceClose:1737	Graphs on the device hasn't been destroyed! Graphs will be deallocated
W: [ncAPI] [    540493] [benchmark_app] ncDeviceClose:1745	Fifos on the device hasn't been destroyed! Fifos will be deallocated
E: [global] [    540493] [Scheduler00Thr] dispatcherEventSend:53	Write failed (header) (err -4) | event XLINK_CLOSE_STREAM_REQ

E: [xLink] [    540493] [Scheduler00Thr] sendEvents:1027	Event sending failed
E: [global] [    540493] [Scheduler00Thr] dispatcherEventSend:53	Write failed (header) (err -4) | event XLINK_CLOSE_STREAM_REQ

E: [xLink] [    540493] [Scheduler00Thr] sendEvents:1027	Event sending failed
E: [global] [    540493] [Scheduler00Thr] dispatcherEventSend:53	Write failed (header) (err -4) | event XLINK_RESET_REQ

E: [xLink] [    540493] [Scheduler00Thr] sendEvents:1027	Event sending failed
E: [ncAPI] [    550493] [benchmark_app] ncDeviceClose:1853	Device didn't appear after reboot
[ ERROR ] Failed to read output from FIFO: NC_ERROR

```

We know that this is not the officially supported method of going from Darknet to TF to IR (https://docs.openvinotoolkit.org/2020.4/openvino_docs_MO_DG_prepare_model_convert_model_tf_specific_Convert_YOLO_From_Tensorflow.html), but our entire training infrastructure is in PyTorch and we have no way of moving to Darknet/TensorFlow.

Since the PyTorch->ONNX->IR model works fine on CPU/GPU, it seems to be some limitation of the VPU plugin and/or hardware with the specific operations/shapes in our model.xml. All the layers seem to be supported according to the documentation, so we're not sure how to proceed/debug this issue.

Here's a ZIP of the model XML and bin which executes correctly on CPU and GPU, but fails on NCS2 with the above error: https://drive.google.com/file/d/1tFpd5-2D8RzGFgn9BBVhOPlc0BHWDYse/view?usp=sharing

Thanks for the help!

**Comment by jgespino** [comment]
Hi @mhkabir 

Thanks for reaching out. As you mentioned, the OpenVINO toolkit officially supports the YOLOv3 model implemented in the DarkNet Framework. The YOLOv3 implementation in PyTorch may be different and may or may not work on all devices. Using the  IE Core [query network](https://docs.openvinotoolkit.org/latest/ie_python_api/classie__api_1_1IECore.html#a50a801745d856a758ab357dd4cc6afc4) function, the following layers are not supported by the MYRIAD Plugin.

```
not_supported ['input.1', 'Constant_82973', 'Constant_108113', 'Constant_108035', 'Constant_107957', 
'440/reciprocal_6009_const', '432/Output_0/Data__const', '425/Cast_115171_const', '422/Cast_115213_const', 
'Multiply_106524', '375/reciprocal_6003_const', '367/Output_0/Data__const', '360/Cast_115187_const', 
'357/Cast_115237_const', 'Multiply_106521', '310/reciprocal_5997_const', '302/Output_0/Data__const', 
'295/Cast_115125_const', '292/Cast_115175_const', 'Multiply_106518']
```

Regards,
Jesus

**Comment by mhkabir** [comment]
Thanks Jesus, good to know that there is way to check the supported layers by plugin. It would be great to have that in the docs somewhere, and also have the framework explicitly warn us when there are unsupported layers present. 

We'll look into modifying the ONNX export to avoid those layers and get back.

**Comment by mhkabir** [comment]
Out of curiosity, why are layers like "Constant" and "Multiply" unsupported? Do you have any suggestions on what kind of substitutions/replacements to make in this case?

**Comment by leoll2** [comment]
Hello @jgespino ,
I work with @mhkabir in the same project. 
After realizing that some layers are not supported by MYRIAD, I decided to exploit the Heterogeneous Plugin, which is reportedly meant to "execute not supported layers on fallback devices like CPU".
Hence, the first thing I did was to load the same model specifying `HETERO:MYRIAD,CPU` as device; as far as I understood, the default policy should automatically execute unsupported layers on cpu. Unfortunately, the inference fails with an error like this:
```
E: [global] [     16057] [dnn_detector] XLinkReadData:165	addEventWithPerf(&event, &opTime) failed with error: 3
E: [ncAPI] [     16058] [dnn_detector] ncFifoReadElem:3265	Packet reading is failed.
terminate called after throwing an instance of 'InferenceEngine::details::InferenceEngineException'
  what():  Failed to read output from FIFO: NC_ERROR
```
So I tried to explicitly (manually) set the affinity of each layer to devices, and here comes the interesting part.
If I pin every single layer to CPU (still under hetero:myriad,cpu,  but de facto ignoring the vpu), I obtain this weird error:
```
terminate called after throwing an instance of 'InferenceEngine::details::InferenceEngineException'
  what():  Network passed to LoadNetwork has affinity assigned, but some layers eg: 
(Name:357, Type: Reshape) were not assigned to any device.
It might happen if you assigned layers amnually and missed some layers or
if you used some automatic assigning mode which decided that these layers are not
supported by any plugin
```
The code I used for setting the affinity looks like this:
```
for (auto const& layer : query_res.supportedLayersMap)
      net_reader_.getNetwork().getLayerByName(layer.first.c_str())->affinity = "CPU";
```
Well, I double-checked the affinity of layer 357 (Reshape) after setting it, and it is CPU as expected; however, for some reason the inference engine thinks that it is "not assigned to any device". I don't know why it complains about this Reshape, it's just some random layer in the middle of the network (also supported by VPU according to docs, but not relevant here).

I performed some more experiments to diagnose the issue, and found these two notable behaviours:
- if I set the affinity of layer 357 (Reshape) to CPU, leaving all the other untouched (i.e. default), the system does NOT complain about "not assigned to device" layers, but still dies with the obscure error messages above
- if I set the affinity of one random layer, e.g. 151 (Conv) to CPU, leaving all the other untouched (i.e. default), the system does complain about "not assigned to device" layers, and dies consequently.

Please let me know if you need more information to debug this issue, and if you know any workaround.

_Note: the NCS2 works fine, it runs other models without problems._

**Comment by mhkabir** [comment]
@jgespino Any updates? This seems like an OpenVINO bug.

**Comment by jgespino** [comment]
Hi @mhkabir @leoll2

I'm checking with the development team to understand if Const and Eltwise (multiply) are supported on MYRIAD or if this is a bug. According to the [Supported  Layers](https://docs.openvinotoolkit.org/latest/openvino_docs_IE_DG_supported_plugins_Supported_Devices.html#supported_layers) documentation these should be supported on MYRIAD. The Hetero plugin should run unsupported layers on the specified fallback device, this may be a bug in your case. 

As I mentioned before, OpenVINO supports the DarkNet Framework implementation of YOLOv3, the ONNX implementation has not been validated with OpenVINO.

Regards,
Jesus

**Comment by jgespino** [comment]
Hi @mhkabir @leoll2 

The dev team is looking into your model and layer softmax named "435" with axis param = 1 and shape {N: 8112, C: 1} (softmax by axis C with size 1)  looks strange as it produces an output tensor filled by 1. Could you confirm if this is the expected behavior?

Regards,
Jesus

**Comment by leoll2** [comment]
> The dev team is looking into your model and layer softmax named "435" with axis param = 1 and shape {N: 8112, C: 1} (softmax by axis C with size 1) looks strange as it produces an output tensor filled by 1. Could you confirm if this is the expected behavior?

Hi @jgespino , thanks for investigating our issue.
In our object detector, that softmax layer is responsible for predicting the class; since we have only one class in our applications, it makes sense that it always predicts, i.e. the network is always 100% sure that the object belongs to that specific class, if it's actually an object (objectness is instead the output of the sigmoid nearby).
Might be redundant, yeah, but not wrong.

Let us know if you discover anything.
Thanks again!
Leonardo

**Comment by jgespino** [comment]
Hi @leoll2 

This should be added in the next software release. You can built OpenVINO from source once the pull request below is completed or make the changes yourself on a local build.
https://github.com/openvinotoolkit/openvino/pull/1977

Regards,
Jesus

**Comment by jgespino** [comment]
Hi @leoll2 @mhkabir 

At this time we are unable to merge a fix to master branch or the next release as the proposed solution could affect other topologies. Instead, could you try reconverting your model by adding `--output 446/Split` to the model optimizer command?

Regards,
Jesus

**Comment by jgespino** [comment]
Hi @leoll2 @mhkabir 

Did you get a chance to try adding `--output 446/Split` to your MO command?

Regards,
Jesus

**Comment by leoll2** [comment]
Hi @jgespino, I tried converting with `--output 446` and I still get the same hardware errors.

**Comment by jgespino** [comment]
@leoll2 Thanks for giving that a try, let me check with the development team and get back to you.

**Comment by jgespino** [comment]
Hi @leoll2 

Quick update, there is a PR that should contain a fix for this issue. Please see https://github.com/openvinotoolkit/openvino/pull/2885.

Regards,
Jesus

**Comment by jgespino** [comment]
@mhkabir @leoll2 Fix has been merged to master branch. Please try build from source and let us know if you have any issues.
